### DISCUSSION TITLE
Discord Bot Technical Specification Review

### DISCUSSION THREAD

**Discussion by Caleb-Cohen** [Together-100Devs/Together#395] [Technical Specification Review]
### Summary

This Technical Specification covers the discord bot, potential features and goals.

### Scope

- A discord bot that can join servers
   - Functionality includes: PMing users that have RSVP'd to events, and messaging in general a certain event is about to start.
- Updates to the Event/User objects to pass the information to discord bot
- Updating UI to allow users to subscribe/unsubscribe from events

### Data Model

Add rsvpList array to all current and future event objects. If a user RSVP's for an event add identifying information to rsvpList array.

### User Stories

User sees an event they are interested in on the together calendar, open the event modal, and click RSVP. When the event is <15 minutes away user will receive a PM from Together mentioning their event is about to start.

Allow users to RSVP for single or all events in a recurring group. Allow users to remove themselves from an RSVP list in the same manner. 

Additionally, the discord bot would post upcoming events to the general chats in 15 minutes intervals. 

### Implementation Details

- Setup discord bot capable of private messaging users & posting to general. Utilize the least amount of privileges as possible. Ideally, nothing more than view channels & send messages
- Receive clearance from Leon/100Devs
- Update DB to store user identifiers in rsvpList array 
- Build out bot functionality in the test server to PM specific users/post about specific events

### APIs

- Bot will need to request DB for events starting soon
- API points for adding and removing users from event object

### Security/Privacy

The biggest concern point is people posting events that violate rules. @InstincDev is working on admin.js to prevent events from not being available unless they are approved. 

### Testing

I am not familiar with testing discord bots, but we'll need to add testing to verify users can RSVP and remove themselves from events.

### Open Questions

What is stored in the rsvpList MongoDB userId, disocordName, or DiscordId?

### Deployment

- concerns about damaging production DB event objects.
- security concerns holds this discord bot behind the admin.js project
- requires Leon's approval before proceeding

**Comment by timmyichen** [comment]
> Additionally, the discord bot would post upcoming events to the general chats in 15 minutes intervals.

"general chats" is pluraled - what channels were you thinking of other than #general?  Additionally, how far in advance will it notify of events? i.e. if the bot is sending a message on the :00, :15, :30, :45 marks, and an event is set to start at :16, when will that event be notified?  If at :15, will that be enough of a heads up?

> The biggest concern point is people posting events that violate rules. @InstincDev is working on admin.js to prevent events from not being available unless they are approved.

I think part of the magic of Together is the immediacy.  Thoughts on adding a flag `verified` to events (in which events are verified via adminjs), and unverified events cannot be RSVP'ed to and will not show up in bot messages?

> I am not familiar with testing discord bots, but we'll need to add testing to verify users can RSVP and remove themselves from events.

Can test the behavior of bots by testing the functions that the bots call via unit tests.  You'd mock out most of the discord.js API calls but you can still ensure that specific discord.js methods are being called.

> What is stored in the rsvpList MongoDB userId, disocordName, or DiscordId?

I'd go with `userId` and allow the scheduled job to fetch the discordName if used as a greeting and discordId for messaging the user itself

As an aside, I'd look into using `toad-scheduler` to do the interval messaging: https://github.com/kibertoad/toad-scheduler

**Reply by InstincDev** [reply]
> The biggest concern point is people posting events that violate rules. @InstincDev is working on admin.js to prevent events from not being available unless they are approved.

Info update: Due to complications with the admin.js library, the admin dashboard is currently  created from scratch.

**Reply by Caleb-Cohen** [reply]
> "general chats" is pluraled - what channels were you thinking of other than #general?

Typo! Just General.

> Additionally, how far in advance will it notify of events? i.e. if the bot is sending a message on the :00, :15, :30, :45 marks, and an event is set to start at :16, when will that event be notified? If at :15, will that be enough of a heads up?

Good point. For simplicity I want the bot to poll the DB without being too intensive. I guess we can have the bot poll every minute and post if there is an event 15 minutes away.

I've also considered standardizing the start times of events to :00, :15, :30, :45 and that would make it easier to poll as well.

> I think part of the magic of Together is the immediacy. Thoughts on adding a flag verified to events (in which events are verified via adminjs), and unverified events cannot be RSVP'ed to and will not show up in bot messages?

I actually like that a lot. 

Thanks for the feedback, I will take a look at toad-scheduler.

**Comment by MatthewBozin** [comment]
> A discord bot that can join servers

I'd recommend scoping the bot to exist in just the 100Devs server for the MVP. If the bot was in multiple servers, we'd have to add some steps to the setup process.

> Updates to the Event/User objects to pass the information to discord bot

Does this refer to the Data Model section? Also might be work changing to "make the data usable by both the Together server and the discord bot"

**Reply by Caleb-Cohen** [reply]
> I'd recommend scoping the bot to exist in just the 100Devs server for the MVP. If the bot was in multiple servers, we'd have to add some steps to the setup process.

Oops typo, its only purpose will be to join 100Devs.

> Does this refer to the Data Model section? Also might be work changing to "make the data usable by both the Together server and the discord bot"

Yes it is. Can you expand on work comment?
